### PR TITLE
Harden remote plugin installs

### PR DESCRIPTION
**Pull request by lbajolet-hashicorp**
Since we're hardening what Packer is able to load locally when it comes to plugins, we need also to harden the installation process a bit.

While testing we noticed some remotes had published their plugins with version mismatches between the tag and the binary.
This was not a problem in the past, as Packer did not care for this, only the binary name was important, and the plugin could be installed without problem.

Nowadays however, since Packer enforces the plugin version reported in the name to be the same as the plugin self-reported version, this makes it impossible for the installed plugin to load anymore in such an instance.

Therefore in order to limit confusion, and so users are able to understand the problem and report it to the plugins with that mismatch, we reject the installations that expose this mismatch, and report it to the user if they cannot install anything else.

### CODE REVIEW

File: packer/plugin-getter/plugins.go
Comment (nywilken) [inline packer/plugin-getter/plugins.go:864]:
```suggestion
							err := fmt.Errorf("release v%s binary reports version %q, which is unsupported. This is likely a problem with the plugin. Try opening an issue on the plugin repository asking them to update the plugin's version information.", version, desc.Version)
```

File: packer/plugin-getter/plugins.go
Comment (lbajolet-hashicorp) [reply packer/plugin-getter/plugins.go:831]: Good call-out! I did add the deletion for the zipfile, but forgot for the plugin binary, that'll be in the next reroll

Comment (lbajolet-hashicorp) [reply packer/plugin-getter/plugins.go:831]: For context: "temporary files may get auto deleted" > depends!
On Linux systems generally `/tmp` is a `tmpfs` volume, which does get wiped at shutdown (backed by virtual memory), on Windows it tends to be stored in the user's homedir under `Appdata`, which is not automatically wiped, so Windows users may experience bloat, moreso than UNIX users.

File: packer/plugin-getter/plugins.go
Comment (nywilken) [inline packer/plugin-getter/plugins.go:194]: This seems duplicated by the lines about. We may be able to drop this.

File: packer/plugin-getter/plugins.go
Comment (nywilken) [inline packer/plugin-getter/plugins.go:827]: For consistency purposes lets drop the use to createTemp for openFile.

File: packer/plugin-getter/plugins.go
Comment (nywilken) [inline packer/plugin-getter/plugins.go:841]: If we could do this in openFile that would be nice. I don't know how this works on Windows. But my understanding on Windows is that if it ends with the `.exe` extension it is flag as executable.

File: packer/plugin-getter/plugins.go
Comment (nywilken) [inline packer/plugin-getter/plugins.go:939]: Let's not make this a pointer..😀

**Review by enolan55 — APPROVED** [review]

**Comment by nywilken** [comment]
So this change started because we enforce the version match when installing via `packer plugins install --path`. I can't stop but wonder if instead of failing on version mismatch (describe vs final binary) if we should instead warn the user of the mismatch and install the plugin as a dev version. 

I say that because for a user wishing to test a new version of a binary, if there is a mismatch, I will have to open an issue on the parent repo and wait. Or I can build the binary with the correct version which then introduces a new step of where someone has to build a binary just to fix the version. Or I can rename the binary to match the version which may be off the next time I install. All options which will need to be documented. IMO any manual intervention will bring CI/CD to halt until addressed. 

By forcing the version mismatch at local binary install I can see folks acting to it sooner. If we default the version to a dev version I also see people continuing to use the binary but can then decide what to do. They may be okay with the binary being a dev version, especially since required_versions supports dev version matching. What do you think?

**Comment by lbajolet-hashicorp** [comment]
> So this change started because we enforce the version match when installing via `packer plugins install --path`. I can't stop but wonder if instead of failing on version mismatch (describe vs final binary) if we should instead warn the user of the mismatch and install the plugin as a dev version.
> 
> I say that because for a user wishing to test a new version of a binary, if there is a mismatch, I will have to open an issue on the parent repo and wait. Or I can build the binary with the correct version which then introduces a new step of where someone has to build a binary just to fix the version. Or I can rename the binary to match the version which may be off the next time I install. All options which will need to be documented. IMO any manual intervention will bring CI/CD to halt until addressed.
> 
> By forcing the version mismatch at local binary install I can see folks acting to it sooner. If we default the version to a dev version I also see people continuing to use the binary but can then decide what to do. They may be okay with the binary being a dev version, especially since required_versions supports dev version matching. What do you think?

Hmmmm, I like the idea honestly. This may indeed be a good bridging solution which may motivate users to report the problem upstream while not completely blocking them.

I'll see what I can do to amend this PR to adopt this behaviour instead.

File: packer/plugin-getter/plugins.go
Comment (sylviamoss) [inline packer/plugin-getter/plugins.go:857]: This code is also adding hardening to the installation of prereleases, right? Should this check be after the check of versions?

Comment (lbajolet-hashicorp) [reply packer/plugin-getter/plugins.go:857]: Pre-release remote installation is not supported at the moment, only local installation is through `packer plugins install --path`.
Though with Wilken's last remark, I guess this may change so that the version remotely installed could be a `dev` one, albeit with a warning that there's a mismatch.
But in any case, describing a constraint with `-dev` in the version, or installing pre-releases through `packer plugins install` without passing in a path is not a supported workflow for the moment.

**Comment by nywilken** [comment]
I added this description to the CHANGELOG as a temporary entry.

```shell
core: Remote plugins installed containing an internal version number that differs from the version
number within the binary name can lead to confusion when tracking Packer plugin version 
information. To help track such deprecates in the plugin version, packer init and packer plugin install have been updated to install such plugins as dev versions (e.g. 1.0.0-dev). 
Users are encouraged to notify plugin maintainers of any version mismatches. https://github.com/hashicorp/packer/pull/12915
```

I think that for remote installation and local installation a version mismatch should not fail warn and treat the binary as a development binary. This will not break user flows who may have this issue now. In the future if we need to we can make it a hard error. Thoughts.

**Comment by lbajolet-hashicorp** [comment]
> hmmm. It looks like the following situation in Packer today errors silently. I tried renaming a binary to have a different version and Packer errors saying the the provision was not found. Looking at the logs I found the message indicating why it was not loaded. But this was not straight forward.
> 
> ```
> Options{PluginDirectory:"/Users/dev/.packer.d/plugins", BinaryInstallationOptions:plugingetter.BinaryInstallatio
> nOptions{APIVersionMajor:"5", APIVersionMinor:"0", OS:"darwin", ARCH:"amd64", Ext:"", Checksummers:[]plugingetter.C
> hecksummer{plugingetter.Checksummer{Type:"sha256", Hash:(*sha256.digest)(0xc0001da180)}}, ReleasesOnly:false}}
> 2024/04/29 11:17:10 plugin "/Users/dev/.packer.d/plugins/github.com/sylviamoss/comment/packer-plugin-comment_v1.
> 0.0-dev_x5.0_darwin_amd64" reported version "1.0.0" while its name implies version "v1.0.0-dev", ignoring
> ```

Ah yes, this is in line with how listing installations worked previously.
Maybe we should print those errors out in the diag instead of just reporting that no plugin was found in cases like these, so something like how `InstallLatest` works, where we stockpile errors and print them out only if the operation failed. 

I'm a bit skeptical of this approach to an extent as it may report too many errors and drown the problem a bit, but this would help in situations like the one you describe.

**Comment by nywilken** [comment]
While I understand the rationale for wanting to include the check. It is adding an additional amount of work and a bit of unknowns that I fear will become a hurdle for users. Our goal for predictable loading was to consolidate plugin loading into a single binary name and directory. This additional change seems a bit premature at this time. Especially since less than a handful of plugins are running into this case. 

Once we add this behavior to an official release it would be hard to retract it if it is relied upon. Seeing as this has never been an issue, let's hold and revisit if and when it becomes a thing. The new HCP Packer version tracking will bring more visibility into the versions being used. 

I suggest we revert the change where we require the version in the binary name to match the version in the describe command as the next step. https://github.com/hashicorp/packer/pull/12888

**Review by nywilken — APPROVED** [review]
Approved once #12953 has been merged in.

**Pull request merged**
